crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -390,20 +391,23 @@ where
     .map_err(|error| anyhow::anyhow!(error))?;
     // TODO: Consider introducing and using global in-memory segment header cache (this comment is
     //  in multiple files)
-    let segment_commitments_cache = Mutex::new(LruCache::new(RECORDS_ROOTS_CACHE_SIZE));
-    let piece_provider = PieceProvider::new(
+    let segment_commitments_cache = Arc::new(Mutex::new(LruCache::new(RECORDS_ROOTS_CACHE_SIZE)));
+    let validator = Some(SegmentCommitmentPieceValidator::new(
         node.clone(),
-        Some(SegmentCommitmentPieceValidator::new(
-            node.clone(),
-            node_client.clone(),
-            kzg.clone(),
-            segment_commitments_cache,
-        )),
-    );
+        node_client.clone(),
+        kzg.clone(),
+        segment_commitments_cache,
+    ));
+
+    let piece_provider = PieceProvider::new(node.clone(), validator.clone());
+
+    let random_walking_piece_provider =
+        RandomWalkingPieceProvider::new(node.clone(), validator.clone());
 
     let piece_getter = Arc::new(FarmerPieceGetter::new(
         node.clone(),
         piece_provider,
+        random_walking_piece_provider,
```

Comment:
`random_walking_piece_provider` makes sense to me, but then `piece_provider` should be renamed to something more specific instead.

On a similar note since `RandomWalkingPieceProvider` gives you a high-level `get_piece` method, `PieceProvider` should probably do that as well (instead right now `FarmerPieceGetter` first extracts peers itself and makes explicit requests to specific peers), but while it does give you `get_piece` and `get_piece_from_peer`, those two methods are doing completely different things.

All of this makes me thinking architecture can be improved significantly. Maybe it is worth improving naming and API of `PieceProvider` and just add another method for random walking there instead of introducing a whole separate data structure?

Reply:
> On a similar note since RandomWalkingPieceProvider gives you a high-level get_piece method, PieceProvider should probably do that as well (instead right now FarmerPieceGetter first extracts peers itself and makes explicit requests to specific peers), but while it does give you get_piece and get_piece_from_peer, those two methods are doing completely different things.

It's not entirely correct. `FarmerPieceGetter` uses both `get_piece` and `get_piece_from_piece` during different stages.

> Maybe it is worth improving naming and API of PieceProvider and just add another method for random walking there instead of introducing a whole separate data structure?

It's probably easier to extend the functionality of the separate structure but it can be merged indeed. Please, confirm that you prefer a merged version.

Reply:
> It's probably easier to extend the functionality of the separate structure but it can be merged indeed. Please, confirm that you prefer a merged version.

I would definitely prefer merged version with methods renamed to something more logical and connected peers collection logic moved from `FarmerPieceGetter` into `PieceProvider` as well. That way we have one provider, better method names and APIs  and we wouldn't need `Node` as a dependency in `FarmerPieceGetter` anymore.

We can even abstract away the fact that we retrieve archival history in two different ways (connected peers and random walk) behind one public method.